python/algos/layers.py

```diff
@@ -370,4 +388,123 @@ def forward(self, inputs):
             pooling_outputs.append(X_i.sum(dim=-1))
         concate_vec = torch.cat(pooling_outputs, dim=-1)
         output = self.fc(concate_vec)
-        return output
\ No newline at end of file
+        return output
+
+# Sequence Attention Layer
+# This code is adapted from github repository:  https://github.com/RUCAIBox/RecBole/blob/master/recbole/model/layers.py 
+class DIEN_DIN_AttLayer(torch.nn.Module):
+    def __init__(self, input_dim, att_hidden_size, att_activation, att_dropout, use_att_bn):
+        super(DIEN_DIN_AttLayer, self).__init__()
+        self.att_mlp_layers = MLPLayer(input_dim = input_dim*4, 
+                                       output_dim = 1, 
+                                       hidden_units = att_hidden_size, 
+                                       hidden_activations = att_activation, 
+                                       final_activation = att_activation,
+                                       dropout_rates = att_dropout, 
+                                       batch_norm = use_att_bn)
+        
+    def forward(self, query, keys, keys_length):
+        batch_size, max_length, dim = keys.size()
+        mask_mat = torch.arange(max_length).view(1, -1)  
+        query = query.repeat(1, max_length)
```

Comment:
Using expand is better than repeat ?

https://discuss.pytorch.org/t/torch-repeat-and-torch-expand-which-to-use/27969